src/charm.py

```diff
@@ -0,0 +1,241 @@
+import ops.main
+import ops.model
+from charms.data_platform_libs.v0.object_storage import (
+    AzureStorageProvides,
+    CredentialRequestedEvent,
+)
+from ops.charm import ActionEvent, ConfigChangedEvent, StartEvent
+from ops.model import ActiveStatus, BlockedStatus
+
+from constants import AZURE_MANDATORY_OPTIONS, AZURE_OPTIONS, AZURE_RELATION_NAME
+
+logger = logging.getLogger(__name__)
+
+
+class ObjectStorageIntegratorCharm(ops.charm.CharmBase):
+    """Charm for object storage integrator service."""
+
+    def __init__(self, *args) -> None:
+        super().__init__(*args)
+        self.azure_provider = AzureStorageProvides(self, AZURE_RELATION_NAME)
+
+        self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+
+        self.framework.observe(self.on.secret_changed, self._on_secret_changed)
+        # self.framework.observe(self.on.secret_changed, self._on_secret_removed)
+        self.framework.observe(self.on.update_status, self._on_update_status)
+
+        self.framework.observe(
+            self.azure_provider.on.credentials_requested, self._on_azure_credentials_requested
+        )
+
+        # actions
+        self.framework.observe(
+            self.on.get_azure_connection_info_action, self.on_get_connection_info_action
+        )
+
+    def get_missing_parameters(self) -> List[str]:
+        """Returns the missing mandatory parameters that are not stored in the config."""
+        missing_options = []
+        for config_option in AZURE_MANDATORY_OPTIONS:
+            if not self.config.get(config_option):
+                missing_options.append(config_option)
+        return missing_options
+
+    def check_and_set_status(self, set_active_if_passed=False) -> bool:
+        """Check for missing parameters and set status of the unit.
+
+        If set_active_if_passed is True, the unit status will be set to Active once
+        it is found there are no missing parameters.
+        """
+        missing_options = self.get_missing_parameters()
+        if missing_options:
+            logger.warning(f"Missing parameters: {missing_options}")
+            self.unit.status = BlockedStatus(f"Missing parameters: {missing_options}")
+            return
+        try:
+            self.decode_secret_key(self.config.get("credentials"))
+        except Exception as e:
+            logger.warning(f"Error in decoding secret: {e}")
+            self.unit.status = BlockedStatus(str(e))
+            return
+
+        if set_active_if_passed:
+            self.unit.status = ActiveStatus()
+
+    def _on_start(self, _: StartEvent) -> None:
+        """Handle the charm startup event."""
+        self.check_and_set_status()
+
+    def _on_update_status(self, event: ops.UpdateStatusEvent):
+        """Handle the update status event."""
+        self.check_and_set_status(set_active_if_passed=True)
+
+    def _on_secret_changed(self, event: ops.SecretChangedEvent):
```

Comment:
This handler will handle the case when user changes the secret that has been provided as value of config option `credentials`. 

For eg,

```
juju config object-storage-integrator credentials=test-cred 

juju update-secret test-cred secret-key=new-value
```

When something like this happens, the charm should fetch the new value of `secret-key` and inform the relation that the secret key has changed.